File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.cpp
class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<int>vis(n,0);
        vector<vector<int>>adj(n);
        for(auto e:edges){
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }
        int count=0;
        for(int i=0;i<n;i++){
            if(vis[i]==0){
                dfs(adj,vis,i);
                count++;
            }
        }
        return count;
    }
    void dfs(vector<vector<int>>&adj, vector<int>&vis, int node){
        if(vis[node]==1){
            return;
        }
        vis[node]=1;
        for(int n:adj[node]){
            if(vis[n]==0){
                dfs(adj,vis,n);
            }
        }
        return;
    }
};

File: 0323-number-of-connected-components-in-an-undirected-graph/0323-number-of-connected-components-in-an-undirected-graph.java
class Solution {
    public int countComponents(int n, int[][] edges) {
        int[] parent = new int[n];
        Arrays.fill(parent, -1);
        int components=n;
        for(int[] edge: edges){
            int x = getParent(parent, edge[0]);
            int y = getParent(parent, edge[1]);
            if(x!=y){
                components--;
                parent[y]= x;
            }
        }
        return components;
    }
    public int getParent(int[] parent, int i){
        if(parent[i] == -1){
            return i;
        }
        return parent[i]=getParent(parent, parent[i]);
    }
}